flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java

```diff
@@ -104,17 +126,24 @@ public FsStateChangelogStorage(
                                 metricGroup,
                                 changelogRegistry)),
                 PREEMPTIVE_PERSIST_THRESHOLD.defaultValue().getBytes(),
-                changelogRegistry);
+                changelogRegistry,
+                localRecoveryConfig);
     }
 
     @VisibleForTesting
     public FsStateChangelogStorage(
             StateChangeUploadScheduler uploader,
             long preEmptivePersistThresholdInBytes,
-            TaskChangelogRegistry changelogRegistry) {
+            TaskChangelogRegistry changelogRegistry,
+            LocalRecoveryConfig localRecoveryConfig) {
         this.preEmptivePersistThresholdInBytes = preEmptivePersistThresholdInBytes;
         this.changelogRegistry = changelogRegistry;
         this.uploader = uploader;
+        this.localRecoveryConfig = localRecoveryConfig;
+        if (localRecoveryConfig.isLocalRecoveryEnabled()) {
+            this.localChangelogRegistry =
+                    new LocalChangelogRegistry(Executors.newSingleThreadExecutor());
```

Comment:
This and `defaultChangelogRegistry` would create two executors for each job, how about passing the `TaskManagerServices.ioExecutor` for this?

Reply:
`TaskManagerServices.ioExecutor` is used for other jobs as well; re-using it might break resource isolation.
It would make sense to share the threads between `TaskChangelogRegistry` and `LocalChangelogRegistry`, but creating two threads per job on TM seems fine too.